File: host/lua/Fnordlicht.lua
#!/usr/bin/lua

--*********************************************************************
-- $Id: $
--
-- written by Timo Boettcher
-- fnordlicht (near) spida (.) net
--
-- (C) 2008,2009 Timo Boettcher
-- This is free software. You may redistribute copies of it under the
-- terms of the GNU General Public License (Version 2)
-- <http://www.gnu.org/licenses/gpl.html>.
-- There is NO WARRANTY, to the extent permitted by law.
--
--*********************************************************************

assert(loadlib("/usr/local/lib/lua/usb/libluausb.so","luaopen_libusb"))()

USB_ID_VENDOR	= "0x16c0"
USB_ID_PRODUCT	= "0x05dc"
USB_MANUFACTURER= "Spida.net"
USB_PRODUCT	= "Fnordlicht"

USBRQ_VERSION			= 0
USBRQ_RESET			= 1
USBRQ_ECHO			= 2
USBRQ_ENUMERATE			= 3
USBRQ_GET_STATUS		= 4
USBRQ_PREPARE_COLOR_PARAM	= 5
USBRQ_PREPARE_FADE_PARAM	= 6
USBRQ_EXEC_SET_COLOR		= 7
USBRQ_EXEC_FADETO_COLOR		= 8

requesttype = 193 --usb.TYPE_VENDOR | usb.RECIP_INTERFACE | usb.ENDPOINT_IN

function getopt(arg, options)
	local tab = {}
	for k, v in ipairs(arg) do
		if string.sub(v, 1, 2) == "--" then
			local x = string.find(v, "=", 1, true)
			if x then
				tab[string.sub(v, 3, x-1)] = string.sub(v, x+1)
			else 
				tab[string.sub(v, 3 )] = true
			end
		elseif string.sub(v, 1, 1) == "-" then
			local y = 2
			local l = string.len(v)
			local jopt
			while (y <= l) do
				jopt = string.sub(v, y, y)
				if string.find(options, jopt, 1, true) then
					if y < l then
						tab[jopt] = string.sub(v, y+1)
						y = l
					else
						tab[jopt] = arg[k + 1]
					end
				else
					tab[jopt] = true
				end
				y = y + 1
			end
		end
	end
	return tab
end

function UsbList()
	local buses=libusb.get_busses()

	local handle
	local manuf
	local prod
	local serial
		
	for dirname, bus in pairs(buses) do
		local devices=libusb.get_devices(bus)
		for filename, device in pairs(devices) do

			local descriptor=libusb.device_descriptor(device)
			-- lua 5.1 knows hexadecimal numbers
--			if descriptor.idVendor == USB_ID_VENDOR and descriptor.idProduct == USB_ID_PRODUCT then
			-- so for lua 5.0 use tonumber instead
			if descriptor.idVendor == tonumber(USB_ID_VENDOR) and descriptor.idProduct == tonumber(USB_ID_PRODUCT) then 
				handle = libusb.open(device)
				manuf = libusb.get_string_simple(handle, 1, 64)
				prod = libusb.get_string_simple(handle, 2, 64)
				serial = libusb.get_string_simple(handle, 3, 64)
				libusb.close(handle)
				if manuf == USB_MANUFACTURER and prod == USB_PRODUCT then 
					print (string.format("Found a Device (VendorId %04x, ProductId %04x, Serial %s) at bus %s",descriptor.idVendor, descriptor.idProduct, serial, dirname))
				end
			end
		end
	end
end

function FindDevice()
	local buses=libusb.get_busses()
	local found_device
		
	for dirname, bus in pairs(buses) do
		local devices=libusb.get_devices(bus)
		for filename, device in pairs(devices) do

			local descriptor=libusb.device_descriptor(device)
			local serial=0
			-- lua 5.1 knows hexadecimal numbers
--			if descriptor.idVendor == USB_ID_VENDOR and descriptor.idProduct == USB_ID_PRODUCT then
			-- so for lua 5.0 use tonumber instead
			if descriptor.idVendor == tonumber(USB_ID_VENDOR) and descriptor.idProduct == tonumber(USB_ID_PRODUCT) then 
				found_device = device
			end
		end
	end
	return found_device
end

function OpenDevice(device)
	local handle
	handle = libusb.open(device)
	return handle
end

function Version(handle, busid, deviceid)
	local buf = "123456"
	result = libusb.control_msg(handle, requesttype, USBRQ_VERSION, busid, deviceid, buf, 6)
	print(string.format("Fnordlicht at 0 has version %d,%d,%d,%d,%d,%d", string.byte(result,1), string.byte(result,2), string.byte(result,3), string.byte(result,4), string.byte(result,5), string.byte(result,6)))
end

function Reset(handle, busid, deviceid)
 	libusb.control_msg(handle, requesttype, USBRQ_RESET, busid, deviceid)
end

-- function Enumerate(busid=0)
-- end

-- function EchoDevice(handle)
-- end

function Status(handle, busid, deviceid)
	local buf = "1234"
	result = libusb.control_msg(handle, requesttype, USBRQ_GET_STATUS, busid, deviceid, buf, 4)
	print(string.format("(%d, %d, %d, %d)", string.byte(result,1), string.byte(result,2), string.byte(result,3), string.byte(result,4)))
end

function SetColor(handle, R, G, B, W, busid, deviceid)
	if 0 <= R and R <256 and 0 <= B and B <256 and 0 <= G and G <256 and 0 <= W and W <256 then
		RG = R * 256 + G
		BW = B * 256 + W
		libusb.control_msg(handle, requesttype, USBRQ_PREPARE_COLOR_PARAM, RG, BW)
		libusb.control_msg(handle, requesttype, USBRQ_EXEC_SET_COLOR, busid, deviceid)
	else
		print "Invalid Color"
	end
end

function FadeColor(handle, R, G, B, W, time, busid, deviceid)
	if 0 <= R and R <256 and 0 <= B and B <256 and 0 <= G and G <256 and 0 <= W and W <256 then
		if 0 <= time and time < 65536 then
			RG = R * 256 + G
			BW = B * 256 + W
			libusb.control_msg(handle, requesttype, USBRQ_PREPARE_COLOR_PARAM, RG, BW)
			libusb.control_msg(handle, requesttype, USBRQ_PREPARE_FADE_PARAM, time, 0)
			libusb.control_msg(handle, requesttype, USBRQ_EXEC_FADETO_COLOR, busid, deviceid)
		else
			print "Invalid Fadetime"
		end
	else
		print "Invalid Color"
	end
end

opts = getopt(arg, "lr")
for k, v in pairs(opts) do
	print(k, v)
end

mode = "s"

if mode == "l" then
	UsbList()
else
	local device
	local handle
	device = FindDevice()
	handle = OpenDevice(device)
	local busid = 0
	local deviceid = 0

	if mode == "r" then
		Reset(handle, busid, deviceid)
	elseif mode == "e" then
	elseif mode == "v" then
		Version(handle, busid, deviceid)
	elseif mode == "t" then
	elseif mode == "s" then
		Status(handle, busid, deviceid)
	elseif mode == "c" then
		R = 1
		B = 10
		G = 80
		W = 255
		SetColor(handle, R, G, B, W, busid, deviceid)
	elseif mode == "f" then
		R = 255
		B = 80
		G = 10
		W = 1
		time = 720
		FadeColor(handle, R, G, B, W, time, busid, deviceid)
	end

end

File: host/ruby/Fnordlicht.rb
#!/usr/bin/ruby -w

#*********************************************************************
# $Id: $
#
# written by Timo Boettcher
# usbnotify (near) spida (.) net
#
# (C) 2008,2009 Timo Boettcher
# This is free software. You may redistribute copies of it under the
# terms of the GNU General Public License (Version 2)
# <http://www.gnu.org/licenses/gpl.html>.
# There is NO WARRANTY, to the extent permitted by law.
#
#*********************************************************************

require 'rubygems'
gem 'ruby-usb'
require 'usb'
require 'optparse'

class Fnordlicht
	USB_ID_VENDOR			= 0x16c0
	USB_ID_PRODUCT			= 0x05dc
	USB_MANUFACTURER		= "Spida.net"
	USB_PRODUCT			= "Fnordlicht

	USBRQ_VERSION			= 0
	USBRQ_RESET			= 1
	USBRQ_ECHO			= 2
	USBRQ_ENUMERATE			= 3
	USBRQ_GET_STATUS		= 4
	USBRQ_PREPARE_COLOR_PARAM	= 5
	USBRQ_PREPARE_FADE_PARAM	= 6
	USBRQ_EXEC_SET_COLOR		= 7
	USBRQ_EXEC_FADETO_COLOR		= 8
	
	REQUESTTYPE = USB::USB_TYPE_VENDOR | USB::USB_RECIP_INTERFACE | USB::USB_ENDPOINT_IN
	INTERFACE = 0
	TIMEOUT = 1000

	def Fnordlicht.list
		::USB.devices.each { |d|
			if d.idVendor == USB_ID_VENDOR and d.idProduct == USB_ID_PRODUCT and d.product == USB_PRODUCT and d.manufacturer == USB_MANUFACTURER
				puts "Found a Device (VendorId 0x%04x, ProductId 0x%04x, Serial %s)" % [d.idVendor, d.idProduct, d.serial_number]
			end
		} || raise("Couldn't find a Device (VendorId %s, ProductId %s)", USB_ID_VENDOR, USB_ID_PRODUCT)
	end

	def initialize
		device = finddevice()
		@handle = opendevice(device)
	end

	def finddevice
		device = ::USB.devices.select { |d|
			d.idVendor == USB_ID_VENDOR and d.idProduct == USB_ID_PRODUCT and d.product == USB_PRODUCT and d.manufacturer == USB_MANUFACTURER
		} || raise("Couldn't find a Device (VendorId %s, ProductId %s)", USB_ID_VENDOR, USB_ID_PRODUCT)

		return device.at(0)
	end

	def opendevice(device)
		handle = device.open
		raise "Unable to obtain a handle to the device." if handle.nil?

		retries = 0
		begin
			error_code = handle.usb_claim_interface(INTERFACE)
			raise unless error_code.nil?
		rescue
			handle.usb_detach_kernel_driver_np(INTERFACE);
			if retries.zero? 
				retries += 1
				retry
			else
				raise "Unable to claim the device interface."
			end
		end

		raise "Unable to set alternative interface." unless handle.set_altinterface(0).nil?
		return handle
	end

	def version(busid, deviceid)
		buffer = "\0" * 6
		ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_VERSION, busid, deviceid, buffer, TIMEOUT)
		return "Fnordlicht at 0 has version (%d, %d, %d, %d, %d, %d)" %buffer.unpack('C*')
	end

	def enumerate()
	end

	def reset()
		buffer = ""
		print "Resetting device at 0"
		ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_RESET, 0, 0, buffer, TIMEOUT)
	end

	def echodevice()
		buffer = "\0" * 4
		0.upto(255).each { |senda| 
			ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_ECHO, senda, senda, buffer, TIMEOUT)
			resultbuffer = buffer.unpack("C*")
			if resultbuffer[0] != senda or resultbuffer[1] != senda or resultbuffer[2] != senda or resultbuffer[3] != senda
				puts "Sent: %d, Received %d %d %d %d\n"% [senda, resultbuffer[0], resultbuffer[1], resultbuffer[2], resultbuffer[3]]
			end
		}
	end

	def status(busid, deviceid)
		buffer = "\0" * 4
		ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_GET_STATUS, busid, deviceid, buffer, TIMEOUT)
		return "(%d, %d, %d, %d)" %buffer.unpack('C*')
	end

	def setcolor(r, g, b, w, busid, deviceid)
		if 0 <= r and r <256 and 0 <= b and b <256 and 0 <= g and g <256 and 0 <= w and w <256
			buffer = ""
			rg = (r<<8)+g
			bw = (b<<8)+w
			ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_PREPARE_COLOR_PARAM, rg, bw, buffer, TIMEOUT)
			ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_EXEC_SET_COLOR, busid, deviceid, buffer, TIMEOUT)
		else
			raise "Invalid Color"
		end
	end

	def fadecolor(r, g, b, w, time, busid, deviceid)
		if 0 <= r and r <256 and 0 <= b and b <256 and 0 <= g and g <256 and 0 <= w and w <256
			if 0 <= time and time < 65536
				buffer = ""
				rg = (r<<8)+g
				bw = (b<<8)+w
				ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_PREPARE_COLOR_PARAM, rg, bw, buffer, TIMEOUT)
				ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_PREPARE_FADE_PARAM, time, 0, buffer, TIMEOUT)
				ret = @handle.usb_control_msg(REQUESTTYPE, USBRQ_EXEC_FADETO_COLOR, busid, deviceid, buffer, TIMEOUT)
			else
				raise "Invalid Fadetime"
			end
		else
			raise "Invalid Color"
		end
	end

end

options = {}
options[:deviceid]=0
options[:busid]=0

parser = OptionParser.new
parser.on('--list',                                      '-l',          'list compatible USB devices')         {options[:mode] = "list"}
parser.on('--device [NUM]',                              '-d', Integer, 'USB device')                          {|options[:deviceid]|}
parser.on('--bus [NUM]',                                 '-b', Integer, 'busid (local=0, rs485=1, ir=2)')      {|options[:busid]|}
parser.on('--reset',                                     '-r',          'reset device')                        {options[:mode] = "reset"}
parser.on('--enumerate',                                 '-e',          'enumerate bus connected to device')   {options[:mode] = "enumerate"}
parser.on('--version',                                   '-v',          'device version')                      {options[:mode] = "version"}
parser.on('--test',                                      '-t',          'USB device echotest')                 {options[:mode] = "test"}
parser.on('--status',                                    '-s',          'display status')                      {options[:mode] = "status"}
parser.on('--id [NUM]',                                  '-i', Integer, 'device to use (0 is local (default)') {|options[:id]|}
parser.on('--color [NUM],[NUM]},[NUM],[NUM]',            '-c', Array,   'color to set (R,G,B,W)')              {|options[:r], options[:g], options[:b], options[:w]| options[:mode]="color"}
parser.on('--fade-color [NUM],[NUM]},[NUM],[NUM],[NUM]', '-f', Array,   'color to fade to (R,G,B,W,dSecs)')    {|options[:r], options[:g], options[:b], options[:w], options[:time]| options[:mode]="fade"}
parser.on('--help',                                      '-h',          'Show this help message.')             { puts parser; exit }

parser.parse ARGV

if options[:mode] == "list"
	Fnordlicht.list()
else
	myFnordlicht = Fnordlicht.new()
	case options[:mode]
	when "reset"
		myFnordlicht.reset()
	when "enumerate"
		puts "not implemented"
	when "version"
		puts myFnordlicht.version(options[:deviceid], options[:busid])
	when "test"
		myFnordlicht.echodevice(options[:deviceid], options[:busid])
	when "status"
		puts myFnordlicht.status(0,0)
	when "color"
		r = options[:r].to_i()
		g = options[:g].to_i()
		b = options[:b].to_i()
		w = options[:w].to_i()
		myFnordlicht.setcolor(r, g, b, w, options[:deviceid], options[:busid])
	when "fade"
		r = options[:r].to_i()
		g = options[:g].to_i()
		b = options[:b].to_i()
		w = options[:w].to_i()
		time = options[:time].to_i()
		myFnordlicht.fadecolor(r, g, b, w, time, options[:deviceid], options[:busid])
	end
end

